dir.c

```diff
@@ -1795,16 +1834,22 @@ static enum path_treatment read_directory_recursive(struct dir_struct *dir,
 			subdir_state =
 				read_directory_recursive(dir, path.buf,
 							 path.len, ud,
-							 check_only, pathspec);
+							 check_only, stop_first_file, pathspec);
 			if (subdir_state > dir_state)
 				dir_state = subdir_state;
 		}
 
 		if (check_only) {
 			/* abort early if maximum state has been reached */
-			if (dir_state == path_untracked) {
-				if (cdir.fdir)
-					add_untracked(untracked, path.buf + baselen);
+			if (dir_state > max_dir_state)
+				dir_state = max_dir_state;
+
+			if (dir_state == max_dir_state) {
+				if (dir_state == path_untracked) {
```

Comment:
The logic should be:

If we have found the `max_dir_state`, then we want to break. If we have found `max_dir_state` and `max_dir_state` is `path_untracked`, then I want to add the file to the `untracked` list (depending on `cdir.fdir`). I am open to suggestions on cleaning up this logic.